jupyterhub/app.py

```diff
@@ -2626,6 +2631,15 @@ def service_from_spec(
             service.orm.server = None
 
         if service.oauth_available:
+            self.log.info(
+                f"Creating service {service.name} with oauth_client_id={service.oauth_client_id}"
+            )
+            if not service.oauth_redirect_uri:
+                # redirect uri has a default value if a URL is configured,
+                # but must be specified explicitly for external services
+                raise ValueError(
+                    f"Service {service.name} has oauth configured, but is missing required oauth_redirect_uri."
+                )
```

Comment:
This is asymetric now, and I think it shouldn't be. We don't error with a nice message if `oauth_redirect_uri` is configured but `oauth_client_id` isn't now, only the other way around.

Reply:
oauth_client_id can't be undefined, since it has a default value. oauth_redirect_uri has a default value unless no URL is configured (i.e. this whole situation can only come up for external but not proxied services).